cou/utils/openstack.py

```diff
@@ -13,118 +13,125 @@
+    [1] https://docs.openstack.org/charm-guide/latest/project/charm-delivery.html
+
+    :return: Ordered dictionary containing the version and the possible Openstack release.
+    :rtype: OrderedDict
+    """
+    charms_types = {
+        "ceph": ["ceph-mon", "ceph-fs", "ceph-radosgw", "ceph-osd"],
+        "swift": ["swift-proxy", "swift-storage"],
+        "nova": ["nova-cloud-controller", "nova-compute"],
+        "ovn": ["ovn-dedicated-chassis", "ovn-central"],
+        "neutron": ["neutron-api", "neutron-gateway"],
+        "manila": ["manila-ganesha"],
+        "horizon": ["openstack-dashboard"],
+    }
+    with open("cou/utils/openstack_lookup.csv") as csv_file:
+        openstack_lookup = OrderedDict()
+        service_column_index = 0
+        minor_column_index = 1
+        micro_column_index = 2
+        version_start_column_index = 3
+        csv_reader = csv.reader(csv_file, delimiter=",")
+        header = next(csv_reader)
+        for row in csv_reader:
+            service_dict: defaultdict[str, Any] = defaultdict(OrderedDict)
+            service = row[service_column_index]
+            minor = row[minor_column_index]
+            micro = row[micro_column_index]
+            for column_index in range(version_start_column_index, len(row), 2):
+                os_version, _ = header[column_index].split("-")
+                min_version = row[column_index]
+                max_version = row[column_index + 1]
+                service_dict[os_version] = Versions(
+                    Version(min_version),
+                    Version(max_version),
+                    True if minor == "TRUE" else False,
+                    True if micro == "TRUE" else False,
+                )
+            openstack_lookup[service] = service_dict
+    for charm_type, charms in charms_types.items():
+        for charm in charms:
+            openstack_lookup[charm] = openstack_lookup[charm_type]
+    return openstack_lookup
+
+
+OPENSTACK_LOOKUP = generate_openstack_lookup()
+
+
+def get_latest_compatible_openstack_codename(
+    charm: str, workload_version: str
+) -> Union[str, None]:
+    """Get the latest Openstack codename based on the charm name and workload version.
+
+    :param charm: Charm name.
+    :type charm: str
+    :param workload_version: Workload version of a charm.
+    :type workload_version: str
+    :return: Return the latest compatible Openstack codename or None if not found.
+    :rtype: Union[str, None]
+    """
+    wl_version = Version(workload_version)
+    possible_os_releases = []
+    if not OPENSTACK_LOOKUP.get(charm):
+        logging.warning(
+            (
+                "Not possible to find a compatible Openstack codename for "
+                "charm: %s with workload_version: %s"
+            ),
+            charm,
+            workload_version,
+        )
+        return None
+    for openstack_release, versions in OPENSTACK_LOOKUP[charm].items():
+        if versions.minor and not versions.micro:
```

Comment:
I think I understand. You are using the inclusive ranges shown in the releases pages, but that's very unwieldy when you cannot for sure know what the last available version is (and it theoretically change in the future due to backports). When you then mix components that span multiple versions, and components whose boundaries are not on a major version it becomes a mess. I think you really need to use exclusive ranges.

Here's another suggestion: store the lookup table directly as python objects, and use a Container to more naturally implement (exclusive) range checks and prevent the whole data structure from becoming unwieldy.

For example:

```
# These are your custom ranges
@dataclass
class VersionRange(Container):
    lower: str
    upper: str

    def __contains__(self, version):
        testver = Version(version)
        lowerver = Version(self.lower)
        upperver = Version(self.upper)
        # note the exclusive range
        return testver >= lowerver and testver < upperver
```

Then you can define your mapping, including possible aliases to make things more compact
```
octopus = VersionRange('15.2', '16')

lookup = {
    'ceph': {
        'ussuri': octopus,
        'victoria': octopus,
    },
    'nova': {
        'ussuri': VersionRange('21', '22'),
        'victoria': VersionRange('22', '23'),
    }
}
```

At this point you can use list comprehensions relatively readably:
```
In [16]: [codename for codename, ver_range in lookup['ceph'].items() if '15.5' in ver_range]
Out[16]: ['ussuri', 'victoria']
```

You _could_ still store all your values in a table for easier updating down the line (adding a new release to the lookup dict would not be fun), though mapping a VersionRange object to a cell (or a pair of cells) would not be very straightforward

Reply:
I liked this idea and I think it can simplify the logic. 
If I understood correctly, the `lower` is considered the min version of a release `N` while `upper` is the min version of the release `N+1`. I'm curious how we can identify the OpenStack code name when we are at the latest possible release.

Reply:
Right, or more precisely (though equivalently) `upper` would be the first incompatible version. Perhaps you can find a better name for it.
In order to handle the latest version of a component we would need to handle an unbound `upper` limit (i.e. the default value for `upper` should be `None` and the return boolean should be `True` as long as `testver >= lowerver`). This was just a small example, feel free to rework and expand as needed.